pageserver/src/tenant/timeline/compaction.rs

```diff
@@ -1814,11 +1879,26 @@ impl Timeline {
             }
             let mut selected_layers = Vec::new();
             drop(gc_info);
+            // Pick all the layers intersect or below the gc_cutoff, get the largest LSN in the selected layers.
+            let Some(max_layer_lsn) = layers
+                .iter_historic_layers()
+                .filter(|desc| desc.get_lsn_range().start <= gc_cutoff)
+                .map(|desc| desc.get_lsn_range().end)
+                .max()
+            else {
+                info!("no layers to compact with gc");
+                return Ok(());
+            };
+            // Then, pick all the layers that are below the max_layer_lsn.
             for desc in layers.iter_historic_layers() {
-                if desc.get_lsn_range().start <= gc_cutoff {
+                if desc.get_lsn_range().end <= max_layer_lsn {
                     selected_layers.push(guard.get_from_desc(&desc));
                 }
             }
+            if selected_layers.is_empty() {
+                info!("no layers to compact with gc");
+                return Ok(());
+            }
```

Comment:
This looks like one of the optimizations to layer selection from the design doc.

They look correct but why is this optimization included in this PR? Seems like an orthogonal change (?)

Reply:
Ohh, no, it's not a selection.

It's a change in behavior; it adds more layers to the selection than before.

Reply:
<img width="1076" alt="image" src="https://github.com/user-attachments/assets/3bc6c7c4-8cc6-451f-9d5b-1f1f5a061e9e">

Before this PR, we would not include C.

With this PR, we do include C.

Why the change?

Reply:
Firstly, this is not a valid layer map if C contains multiple keys.

The change is made to consider the result of the legacy compaction, where it will split a delta layer containing a single key into multiple layers on the LSN horizon. To avoid creating delta layers overlapping with original delta layers in this case, we should take all layers below some LSN into the compaction process, instead of only the ones intersecting with the GC horizon.

Reply:
>  To avoid creating delta layers overlapping with original delta layers in this case, we should take all layers below some LSN into the compaction process, instead of only the ones intersecting with the GC horizon.

Ahh, makes sense. Maybe worth a comment? Feel free to resolve

Reply:
added a comment in the codebase